### PR TITLE
Add fan curve configuration

### PR DESCRIPTION
**Pull request by JackDoan**
- [x] Needs #11 to merge
- [x] built on top of #13,. merge that first too
- [ ] need to add support for curve "start temp"
- [ ] squash/rebase

What do you guys think of my approach? I tried to keep the macro-magic to a minimum, but there's 255 attributes to get added and conditionally displayed, and this was the best/simplest I could come up with, but it's so _ugly_. I'd really appreciate some other ideas.

I considered dynamically creating the attributes in `aqc_probe`, but I couldn't seem to find another module that actually did that without using the raw sysfs API, which _seems_ discouraged. There's also no other driver with this many fan-curve config points :/

### CODE REVIEW

File: aquacomputer_d5next.c [hunk 1]
```diff
@@ -765,14 +1204,50 @@ static int aqc_probe(struct hid_device *hdev, const struct hid_device_id *id)
 		ret = -ENOMEM;
 		goto fail_and_close;
 	}
-
+	
 	mutex_init(&priv->mutex);
 
+	if (priv->num_fans > 0) {
+		priv->groups[1] = &aqc_curve_pt_1_pwm_group;
```
Comment (leoratte) [inline aquacomputer_d5next.c:1211]: The groups should be filled starting from 0. In the current state groups[0] is NULL leading to everything after that being ignored

File: aquacomputer_d5next.c [hunk 2]
```diff
@@ -654,6 +691,413 @@ static void aqc_debugfs_init(struct aqc_data *priv)
...
+	if (kstrtol(buf, 10, &val) < 0)
+		return -EINVAL;
+
+	val = val / 10;
+
+	return aqc_fan_curve_store(dev, attr, AQC_FAN_CTRL_CURVE_TEMPS_OFFSET, val);
+}
+
+
+static SENSOR_DEVICE_ATTR_2_RW(pwm0_auto_point1_pwm, aqc_fan_curve_pwm, 0, 1);
```
Comment (leoratte) [inline aquacomputer_d5next.c:772]: Sysfs starts pwm1 so we should do that here. 
idx should start from 0 to simplify offset calculation.
This line would be:
static SENSOR_DEVICE_ATTR_2_RW(pwm1_auto_point1_pwm, aqc_fan_curve_pwm, 0, 0);

File: aquacomputer_d5next.c [hunk 3]
```diff
@@ -654,6 +691,413 @@ static void aqc_debugfs_init(struct aqc_data *priv)
...
+
+static int aqc_fan_curve_store(struct device *dev, struct device_attribute *attr, int offset, long val)
+{
+	struct sensor_device_attribute_2 *sensor_attr = to_sensor_dev_attr_2(attr);
+	struct aqc_data *priv = dev_get_drvdata(dev);
+	int channel = sensor_attr->nr;
+	int idx = sensor_attr->index;
+	u16 to_set;
+
+	if (channel > priv->num_fans)
```
Comment (leoratte) [inline aquacomputer_d5next.c:737]: channel starts at 0 so the highest channel is num_fans - 1

File: aquacomputer_d5next.c [hunk 4]
```diff
@@ -654,6 +691,413 @@ static void aqc_debugfs_init(struct aqc_data *priv)
...
+
+static int aqc_fan_curve_show(struct device *dev, struct device_attribute *attr, int offset)
+{
+	struct sensor_device_attribute_2 *sensor_attr = to_sensor_dev_attr_2(attr);
+	struct aqc_data *priv = dev_get_drvdata(dev);
+	int channel = sensor_attr->nr;
+	int idx = sensor_attr->index;
+	u16 to_get;
+
+	if (channel > priv->num_fans)
```
Comment (leoratte) [inline aquacomputer_d5next.c:703]: same problem with channel number checking

File: aquacomputer_d5next.c [hunk 5]
```diff
@@ -654,6 +691,413 @@ static void aqc_debugfs_init(struct aqc_data *priv)
...
+{
+	struct sensor_device_attribute_2 *sensor_attr = to_sensor_dev_attr_2(attr);
+	struct aqc_data *priv = dev_get_drvdata(dev);
+	int channel = sensor_attr->nr;
+	int idx = sensor_attr->index;
+	u16 to_set;
+
+	if (channel > priv->num_fans)
+		return -ENOENT;
+	if ((idx > NUM_CTRL_CURVE_POINTS) || idx < 0)
```
Comment (leoratte) [inline aquacomputer_d5next.c:739]: idx should go only up to 15, should check for: idx >= NUM_CTRL_CURVE_POINTS

File: aquacomputer_d5next.c [hunk 6]
```diff
@@ -654,6 +691,413 @@ static void aqc_debugfs_init(struct aqc_data *priv)
...
+{
+	struct sensor_device_attribute_2 *sensor_attr = to_sensor_dev_attr_2(attr);
+	struct aqc_data *priv = dev_get_drvdata(dev);
+	int channel = sensor_attr->nr;
+	int idx = sensor_attr->index;
+	u16 to_get;
+
+	if (channel > priv->num_fans)
+		return -ENOENT;
+	if ((idx > NUM_CTRL_CURVE_POINTS) || idx < 0)
```
Comment (leoratte) [inline aquacomputer_d5next.c:705]: same problem with idx boundary checking

File: aquacomputer_d5next.c [hunk 7]
```diff
@@ -65,26 +66,62 @@ static u8 secondary_ctrl_report[] = {
...
 #define AQC_FAN_SPEED_OFFSET		0x08
 
+#define NUM_CTRL_CURVE_POINTS 0x10
+
+#define AQC_FAN_CTRL_PWM_OFFSET 		0x01
+#define AQC_FAN_CTRL_SRC_OFFSET 		0x03
+#define AQC_FAN_CTRL_PID_OFFSET 		0x05
+#define AQC_FAN_CTRL_CURVE_START_OFFSET 	0x13
+#define AQC_FAN_CTRL_CURVE_TEMPS_OFFSET 	0x15
+#define AQC_FAN_CTRL_CURVE_PWMS_OFFSET 		0x25
```
Comment (leoratte) [inline aquacomputer_d5next.c:76]: The pwm values start from 0x35. Should be:
#define AQC_FAN_CTRL_CURVE_PWMS_OFFSET 		0x35

**Review by leoratte — COMMENTED** [review]
Although I don't any ideas to make this prettier, I left some comments to make the code working as i didn't run for me.

**Comment by JackDoan** [comment]
Ah tragic -- I tried to separate out some other code I was working on and I
broke it because I didn't re-test :(

File: aquacomputer_d5next.c (hunk 1 above)
Comment (JackDoan) [reply aquacomputer_d5next.c]: fixed in f65a7cd3!

File: aquacomputer_d5next.c (hunk 2 above)
Comment (JackDoan) [reply aquacomputer_d5next.c]: also fixed in f65a7cd3

File: aquacomputer_d5next.c (hunk 3 above)
Comment (JackDoan) [reply aquacomputer_d5next.c]: fixed in 0be96963, thanks!

File: aquacomputer_d5next.c (hunk 4 above)
Comment (JackDoan) [reply aquacomputer_d5next.c]: fixed in 0be96963, thanks!

File: aquacomputer_d5next.c (hunk 7 above)
Comment (JackDoan) [reply aquacomputer_d5next.c:76]: Are you sure? I'm reading values that seem to be correct on my D5 with this offset.

File: aquacomputer_d5next.c (hunk 6 above)
Comment (JackDoan) [reply aquacomputer_d5next.c]: fixed in 9fb548e6

File: aquacomputer_d5next.c (hunk 5 above)
Comment (JackDoan) [reply aquacomputer_d5next.c]: fixed in 9fb548e6

File: aquacomputer_d5next.c (hunk 7 above)
Comment (JackDoan) [reply aquacomputer_d5next.c]: fixed in c5a55b66

can check with `for x in $(ls hwmon4/pwm2_auto_point*_pwm); do printf "%s->%s\n" $x $(cat $x); done`

File: aquacomputer_d5next.c [hunk 8]
```diff
@@ -654,6 +691,451 @@ static void aqc_debugfs_init(struct aqc_data *priv)
...
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point1_pwm, aqc_fan_curve_pwm, 4, 0);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point2_pwm, aqc_fan_curve_pwm, 4, 1);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point3_pwm, aqc_fan_curve_pwm, 4, 2);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point4_pwm, aqc_fan_curve_pwm, 4, 3);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point5_pwm, aqc_fan_curve_pwm, 4, 4);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point6_pwm, aqc_fan_curve_pwm, 4, 5);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point7_pwm, aqc_fan_curve_pwm, 4, 6);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point8_pwm, aqc_fan_curve_pwm, 4, 7);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point9_pwm, aqc_fan_curve_pwm, 4, 8);
+static SENSOR_DEVICE_ATTR_2_RW(pwm5_auto_point10_pwm, aqc_fan_curve_pwm, 4, 10);
```
Comment (leoratte) [inline aquacomputer_d5next.c:917]: you missed this one.
id should be 9

**Review by leoratte — COMMENTED** [review]
Seems fine now. Only one small mistake left.

File: aquacomputer_d5next.c (hunk 8 above)
Comment (JackDoan) [reply aquacomputer_d5next.c]: thank you very much, I should sleep more :heart:

**Review by leoratte — APPROVED** [review]

**Pull request closed without merging**